File: php/DatabaseHelper.php
<?php

class DatabaseHelper
{
    // Since the connection details are constant, define them as const
    // We can refer to constants like e.g. DatabaseHelper::username
    
	const username = 'a01576524'; // use a + your matriculation number
    const password = 'dbs20'; // use your oracle db password
	const con_string = 'oracle-lab.cs.univie.ac.at:1521/lab';  //on almighty "lab" is sufficient

    // Since we need only one connection object, it can be stored in a member variable.
    // $conn is set in the constructor.
    protected $conn;
    public $shopname;

    // Create connection in the constructor
    public function __construct()
    {
        try {
            // Create connection with the command oci_connect(String(username), String(password), String(connection_string))
            // The @ sign avoids the output of warnings
            // It could be helpful to use the function without the @ symbol during developing process
            $this->conn = @oci_connect(
                DatabaseHelper::username,
                DatabaseHelper::password,
                DatabaseHelper::con_string
            );

            //check if the connection object is != null
            if (!$this->conn) {
                // die(String(message)): stop PHP script and output message:
                die("DB error: Connection can't be established!");
            }
            
            $this->shopname = $this->getShopName(); 

        } catch (Exception $e) {
            die("DB error: {$e->getMessage()}");
        }
    }

    // Used to clean up
    public function __destruct()
    {
        // clean up
        @oci_close($this->conn);
    }

    // function to Set public shopname 
    private function getShopName(){
        $sql = "select shopname from onlineshop fetch next 1 rows only";
        $statement = @oci_parse($this->conn, $sql);
        @oci_execute($statement);
        @oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);
        @oci_free_statement($statement);
        return $res[0]["SHOPNAME"];
    }

    // function to return 2-D array of query result set arranged by rows 
    public function query($sql){
        $statement = @oci_parse($this->conn, $sql);
        @oci_execute($statement);
        @oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);
        @oci_free_statement($statement);
        return $res;  
    }

    public function delete($table,$pkid){
        $sql = "delete from $table where " . $this->getPK($table) . " = $pkid";
        $statement = @oci_parse($this->conn, $sql);
        $ex = @oci_execute($statement);
        if($ex){
            @oci_commit($this->conn);
            return "OK";
        } else {
            return @oci_error($statement)['message'];
        } 
        @oci_free_statement($statement);        
    }

    public function update($sql){
        $statement = @oci_parse($this->conn, $sql);
        $ex = @oci_execute($statement);
        if($ex){
            @oci_commit($this->conn);
            return "OK";
        } else {
            return @oci_error($statement)['message'];
        } 
        @oci_free_statement($statement);        
    }


    public function getPK($table){
        $pks = array("customer" => "customerid", 
                     "category" => "categoryid",
                     "product" => "productid",
                     "orders" => "orderid",
                     "cart" => "cartid",
                     "shipment" => "shipmentid");
        return($pks[$table]);
    }


    public function getCustomerList($toFind=""){
        $sql = "select * from customer";
        if ($toFind!="") $sql .= " where upper(customername) like '%" . strtoupper($toFind) . "%'";
        $sql .= ' order by 1';
        $statement = @oci_parse($this->conn, $sql);
        @oci_execute($statement);
        @oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);
        @oci_free_statement($statement);
        return $res;
    }

    public function getCategoryList($toFind=""){
        $sql = "select * from category";
        if ($toFind!="") $sql .= " where upper(categoryname) like '%" . strtoupper($toFind) . "%'";
        $sql .= ' order by 1';
        $statement = @oci_parse($this->conn, $sql);
        @oci_execute($statement);
        @oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);
        @oci_free_statement($statement);
        return $res;
    }

    public function getProductList($toFind=""){
        $sql = "select product.productid, product.productname, category.categoryname category, product.price, product.quantity 
        from product left join category on product.categoryid = category.categoryid";
        if ($toFind!="") $sql .= " where upper(product.productname) like '%" . strtoupper($toFind) . "%'";
        $sql .= ' order by 1';
        $statement = @oci_parse($this->conn, $sql);
        @oci_execute($statement);
        @oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);
        @oci_free_statement($statement);
        return $res;
    }

    public function getOrdersList($toFind=""){
        $sql = "select orders.orderid, orders.orderdate, orders.customerid, customer.customername, orders.ordersize, orders.ordercost
        from orders left join customer on orders.customerid = customer.customerid";

        if ($toFind!="") {
            if (is_numeric($toFind)){
                $sql .= " where orderid = $toFind ";
            } else {
                $sql .= " where upper(customer.customername) like '%" . strtoupper($toFind) . "%'";
            }
        }
        
        $sql .= ' order by 1';
        $statement = @oci_parse($this->conn, $sql);
        @oci_execute($statement);
        @oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);
        @oci_free_statement($statement);
        return $res;
    }

    public function getCartList($toFind=""){
        $sql = "select cart.cartid, cart.orderid, substr(product.productname,1,50) productname, cart.price, cart.quantity, cart.quantity * cart.price amount 
        from cart left join product on cart.productid = product.productid";
        if ($toFind!="") $sql .= " where orderid=$toFind";
        $sql .= ' order by 1';
        $statement = @oci_parse($this->conn, $sql);
        @oci_execute($statement);
        @oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);
        @oci_free_statement($statement);
        return $res;
    }

    public function getShipmentList($toFind=""){
        $sql = "select * from shipment";
        if ($toFind!="") $sql .= " where shipmentid='$toFind' or orderid='$toFind' ";
        $sql .= ' order by 1';
        $statement = @oci_parse($this->conn, $sql);
        @oci_execute($statement);
        @oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);
        @oci_free_statement($statement);
        return $res;
    }

    // This function creates and executes a SQL select statement and returns an array as the result
    // 2-dimensional array: the result array contains nested arrays (each contains the data of a single row)
    public function selectFromCustomerWhere($toFind){
        // Define the sql statement string
        // Notice that the parameters $person_id, $surname, $name in the 'WHERE' clause
        $sql = "SELECT * FROM customer
            WHERE upper(productname) LIKE upper('%{$toFind}%')
            ORDER BY customerid ASC ";
        echo $sql;

        // oci_parse(...) prepares the Oracle statement for execution
        // notice the reference to the class variable $this->conn (set in the constructor)
        $statement = @oci_parse($this->conn, $sql);

        // Executes the statement
        @oci_execute($statement);

        // Fetches multiple rows from a query into a two-dimensional array
        // Parameters of oci_fetch_all:
        //   $statement: must be executed before
        //   $res: will hold the result after the execution of oci_fetch_all
        //   $skip: it's null because we don't need to skip rows
        //   $maxrows: it's null because we want to fetch all rows
        //   $flag: defines how the result is structured: 'by rows' or 'by columns'
        //      OCI_FETCHSTATEMENT_BY_ROW (The outer array will contain one sub-array per query row)
        //      OCI_FETCHSTATEMENT_BY_COLUMN (The outer array will contain one sub-array per query column. This is the default.)
        @oci_fetch_all($statement, $res, 0, 0, OCI_FETCHSTATEMENT_BY_ROW);

        //clean up;
        @oci_free_statement($statement);

        return $res;
    }
	

    // This function creates and executes a SQL insert statement and returns true or false
    public function insertIntoPerson($name, $surname)
    {
        $sql = "INSERT INTO PERSON (NAME, SURNAME) VALUES ('{$name}', '{$surname}')";

        $statement = @oci_parse($this->conn, $sql);
        $success = @oci_execute($statement) && @oci_commit($this->conn);
        @oci_free_statement($statement);
        return $success;
    }

    // Using a Procedure
    // This function uses a SQL procedure to delete a person and returns an errorcode (&errorcode == 1 : OK)
    public function deletePerson($person_id)
    {
        // It is not necessary to assign the output variable,
        // but to be sure that the $errorcode differs after the execution of our procedure we do it anyway
        $errorcode = 0;

        // In our case the procedure P_DELETE_PERSON takes two parameters:
        //  1. person_id (IN parameter)
        //  2. error_code (OUT parameter)

        // The SQL string
        $sql = 'BEGIN P_DELETE_PERSON(:person_id, :errorcode); END;';
        $statement = @oci_parse($this->conn, $sql);

        //  Bind the parameters
        @oci_bind_by_name($statement, ':person_id', $person_id);
        @oci_bind_by_name($statement, ':errorcode', $errorcode);

        // Execute Statement
        @oci_execute($statement);

        //Note: Since we execute COMMIT in our procedure, we don't need to commit it here.
        //@oci_commit($statement); //not necessary

        //Clean Up
        @oci_free_statement($statement);

        //$errorcode == 1 => success
        //$errorcode != 1 => Oracle SQL related errorcode;
        return $errorcode;
    }

    // NOT IN USE - ALTERNATIVE to a simple insert (method return person_id)
    // using a Procedure to add a Person -> the Id of the currently added Person is return (otherwise false)
    public function addPerson($name, $surname)
    {
        $person_id = -1;
        $sql = 'BEGIN P_ADD_PERSON(:name, :surname, :person_id); END;';
        $statement = @oci_parse($this->conn, $sql);

        @oci_bind_by_name($statement, ':name', $name);
        @oci_bind_by_name($statement, ':surname', $surname);
        @oci_bind_by_name($statement, ':person_id', $person_id);


        if (!@oci_execute($statement)) {
            @oci_commit($this->conn);
        }
        @oci_free_statement($statement);

        return $person_id;
    }
}

File: java/src/DatabaseHelper.java
//Database Systems (Module IDS) 

import java.sql.*;
import java.util.ArrayList;

// The DatabaseHelper class encapsulates the communication with our database
class DatabaseHelper {
    // Database connection info
   private static final String DB_CONNECTION_URL = "jdbc:oracle:thin:@oracle-lab.cs.univie.ac.at:1521:lab";
   private static final String USER = "a01576524"; //TODO: use a + matriculation number
   private static final String PASS = "dbs20"; //TODO: use your password (default: dbs19)


    // The name of the class loaded from the ojdbc14.jar driver file
    //private static final String CLASSNAME = "oracle.jdbc.driver.OracleDriver";

    // We need only one Connection and one Statement during the execution => class variable
    private static Statement stmt;
    private static Connection con;

    private String sql;

    //CREATE CONNECTION
    DatabaseHelper() {
        try {
            // establish connection to database
            con = DriverManager.getConnection(DB_CONNECTION_URL, USER, PASS);
            stmt = con.createStatement();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }



    //DROP ALL
    void dropAll(){
        ArrayList<String> tables = new ArrayList<String>();
        tables.add("shipment");
        tables.add("cart");
        tables.add("orders");
        tables.add("product");
        tables.add("customer");
        tables.add("onlineshop_category");
        tables.add("category");
        tables.add("onlineshop");

        for (String table : tables){
            try {
                sql = "drop table " + table + " cascade constraints";
                stmt.execute(sql);
            } catch (Exception e){ }
        }
    }

    void createTables(){
        try {
            sql = "create table onlineshop("
                    + " shopid			integer not null,"
                    + " Shopname	    varchar2(50),"
                    + " website			varchar2(100),"
                    + " constraint pk_onlineshop primary key (shopid)"
                    + " )";
            stmt.execute(sql);

            sql = "create table category("
                    + " categoryid		integer not null,"
                    + " categoryname    varchar2(50),"
                    + " target_market   varchar2(50),"
                    + " constraint pk_category primary key (categoryid)"
                    + " )";
            stmt.execute(sql);

            sql = "create table onlineshop_category("
                    + " shopid			integer not null,"
                    + " categoryid		integer not null,"
                    + " constraint pk_onlineshop_category primary key (shopid, categoryid)"
                    + " )";
            stmt.execute(sql);

            sql = "create table product("
                    + " productid		integer not null,"
                    + " categoryid		integer not null,"
                    + " productname		varchar2(100) not null,"
                    + " quantity		decimal(8,2) not null,"
                    + " price			decimal(8,2) not null,"
                    + " constraint pk_product primary key (productid),"
                    + " constraint ch_quantity check (quantity >1),"
                    + " constraint ch_price check (price >1)"
                    + " )";
            stmt.execute(sql);

            sql = "create table customer("
                    + " customerid      integer not null,"
                    + " customername    varchar2(50),"
                    + " date_of_birth   date ,"
                    + " email   		varchar2(50),"
                    + " address 		varchar2(250),"
                    + " influencerid 	integer,"
                    + " constraint pk_customer primary key (customerid)"
                    + " )";
            stmt.execute(sql);

            sql = "create table orders("
                    + " orderid			integer not null,"
                    + " customerid		integer not null,"
                    + " orderdate		date not null,"
                    + " address			varchar2(250),"
                    + " ordersize		decimal(8,2),"
                    + " ordercost		decimal(8,2),"
                    + " constraint pk_orders primary key (orderid)"
                    + " )";
            stmt.execute(sql);

            sql = "create table cart("
                    + " cartid			integer not null,"
                    + " orderid			integer not null,"
                    + " productid		integer not null,"
                    + " quantity		decimal(8,2),"
                    + " price			decimal(8,2),"
                    + " constraint pk_cart primary key (cartid),"
                    + " constraint ch_quantity_cart CHECK (quantity > 0) ,"
                    + " constraint ch_price_cart CHECK (price > 0)"
                    + " )";
            stmt.execute(sql);

            sql = "create table shipment ("
                    + " shipmentid		integer not null,"
                    + " orderid			integer not null,"
                    + " deliver_time	date,"
                    + " deliver_by		varchar2(50),"
                    + " warehouse		varchar2(50),"
                    + " shipmentsize	decimal(8,2),"
                    + " shipmentcost	decimal(8,2),"
                    + " constraint pk_shipment primary key (shipmentid)"
                    + " )";
            stmt.execute(sql);

        } catch (Exception e){
            System.out.println(sql);
            System.err.println("Error at: createTables\nmessage: " + e.getMessage());
        }
    }

    void createForeignKeys(){
        try {
            sql = "alter table onlineshop_category add constraint fk_onlineshop_categoryonlineshop foreign key (shopid) references onlineshop (shopid)";
            stmt.execute(sql);
            sql = "alter table onlineshop_category add constraint fk_onlineshop_categorycategory foreign key (categoryid) references category (categoryid)";
            stmt.execute(sql);
            sql = "alter table product add constraint fk_product_category foreign key (categoryid) references category (categoryid)";
            stmt.execute(sql);
            sql = "alter table customer add constraint fk_customer foreign key (influencerid) references customer (customerid)";
            stmt.execute(sql);
            sql = "alter table orders add constraint fk_orders_customer foreign key (customerid) references customer (customerid)";
            stmt.execute(sql);
            sql = "alter table cart add constraint fk_cart_orders foreign key (orderid) references orders (orderid)";
            stmt.execute(sql);
            sql = "alter table cart add constraint fk_cart_prodcut foreign key (productid) references product (productid)";
            stmt.execute(sql);
            sql = "alter table shipment add constraint fk_shipment_orders foreign key (orderid) references orders(orderid)";
            stmt.execute(sql);
        } catch (Exception e){
            System.out.println("\n" + sql);
            System.err.println("Error at: createTables\nmessage: " + e.getMessage());
        }
    }

    //INSERT INTO ONLINESHOP
    void insertIntoOnlineShop(Integer ShopId, String ShopName, String Website) {
        try {
            sql = "INSERT INTO OnlineShop (ShopId, ShopName, Website) VALUES (" +
                    ShopId +
                    ",'" +  ShopName + "'" +
                    ",'" +  Website + "'" +
                    ")";
            stmt.execute(sql);

        } catch (Exception e) {
            System.out.println(sql);
            System.err.println("Error at: insertIntoOnlineShop\nmessage: " + e.getMessage());
        }
    }

    //INSERT INTO CATEGORY
    void insertIntoCategory(Integer CategoryId, String CategoryName, String Target_Market) {
        try {
            sql = "INSERT INTO Category (CategoryId, CategoryName, Target_Market) VALUES (" +
                    CategoryId +
                    ",'" +  CategoryName + "'" +
                    ",'" +  Target_Market + "'" +
                    ")";
            stmt.execute(sql);
        } catch (Exception e) {
            System.out.println(sql);
            System.err.println("Error at: insertIntoCategory\nmessage: " + e.getMessage());
        }
    }

    //INSERT INTO ONLINESHOP_CATEGORY
    void insertIntoOnlineShopCategory(Integer Shopid, Integer CategoryId) {
        try {
            sql = "INSERT INTO OnlineShop_Category (ShopId, CategoryId) VALUES (" +
                    Shopid +
                    "," +  CategoryId +
                    ")";
            stmt.execute(sql);
        } catch (Exception e) {
            System.out.println(sql);
            System.err.println("Error at: insertIntoOnlineShopCategory\nmessage: " + e.getMessage());
        }
    }

    //INSERT INTO CUSTOMER
    void insertIntoCustomer(Integer CustomerId, String CustomerName, String Date_of_Birth, String Email, String Address, String InfluencerId) {
        String OptionalInfluencer = InfluencerId;
        try {
            if(OptionalInfluencer.equals("")) OptionalInfluencer="Null";
            String sql = "INSERT INTO Customer (CustomerId, CustomerName, Date_of_Birth, Email, Address, InfluencerId) VALUES (" +
                    CustomerId +
                    ",'" + CustomerName + "'" +
                    ",'" + Date_of_Birth + "'" +
                    ",'" + Email + "'" +
                    ",'" + Address + "'" +
                    "," +  OptionalInfluencer +
                    ")";
            stmt.execute(sql);
        } catch (Exception e) {
            System.err.println("Error at: insertIntoCustomer\nmessage: " + e.getMessage());
        }
    }

    //INSERT INTO PRODUCT
    void insertIntoProduct(Integer ProductId, Integer CategoryId, String ProductName, Float Quantity, Float Price) {
        try {
            sql = "INSERT INTO PRODUCT (ProductId, CategoryId, ProductName, Quantity, Price) VALUES (" +
                    ProductId +
                    "," +  CategoryId +
                    ",'" + ProductName + "'" +
                    "," +  Quantity +
                    "," +  Price +
                    ")";
            stmt.execute(sql);
        } catch (Exception e) {
            System.out.println(sql);
            System.err.println("Error at: insertIntoProduct\nmessage: " + e.getMessage());
        }
    }

    //INSERT INTO ORDERS
    void insertIntoOrders(Integer OrderId, Integer CustomerId, String OrderDate, String Address, Float OrderSize, Float OrderCost) {
        try {
            sql = "INSERT INTO Orders (OrderId, CustomerId, OrderDate, Address, OrderSize, OrderCost) VALUES (" +
                    OrderId +
                    "," + CustomerId +
                    ",'" + OrderDate + "'" +
                    ",'" + Address + "'" +
                    "," + OrderSize +
                    "," + OrderCost +
                    ")";
            stmt.execute(sql);
        } catch (Exception e) {
            System.out.println(sql);
            System.err.println("Error at: insertIntoOrders\nmessage: " + e.getMessage());
        }
    }

    //INSERT INTO CART
    void insertIntoCart(Integer CartId, Integer OrderId, Integer ProductId, Float Quantity, Float Price) {
        try {
            sql = "INSERT INTO Cart (CartId, OrderId, ProductId, Quantity, Price) VALUES (" +
                    CartId +
                    "," + OrderId +
                    "," +  ProductId +
                    "," +  Quantity +
                    "," +  Price +
                    ")";
            stmt.execute(sql);
        } catch (Exception e) {
            System.out.println(sql);
            System.err.println("Error at: insertIntoCart\nmessage: " + e.getMessage());
        }
    }

    //INSERT INTO SHIPMENT
    void insertIntoShipment(Integer ShipmentId, Integer OrderId, String Deliver_Time, String Deliver_By, String Warehouse, Float ShipmentSize, Float ShipmentCost) {
        try {
            sql = "INSERT INTO Shipment (ShipmentId, OrderId, Deliver_Time, Deliver_By, Warehouse, ShipmentSize, ShipmentCost) VALUES (" +
                    ShipmentId +
                    ","  + OrderId +
                    ",'" + Deliver_Time + "'" +
                    ",'" + Deliver_By + "'" +
                    ",'" + Warehouse + "'" +
                    ","  + ShipmentSize +
                    ","  + ShipmentCost +
                    ")";
            stmt.execute(sql);
        } catch (Exception e) {
            System.out.println(sql);
            System.err.println("Error at: insertIntoShipment\nmessage: " + e.getMessage());
        }
    }

    public void close()  {
        try {
            stmt.close(); //clean up
            con.close();
        } catch (Exception ignored) {
        }
    }
}